sql/lease.go

```diff
@@ -0,0 +1,551 @@
+	// nil if there is no lease acquisition in progress for the table. If
+	// non-nil, the channel will be closed when lease acquisition completes.
+	acquiring chan struct{}
+}
+
+func (t *tableState) acquire(version uint32, store LeaseStore) (*LeaseState, error) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	for {
+		s := t.active.findNewest(version)
+		if s != nil {
+			if version != 0 && s != t.active.findNewest(0) {
+				// If a lease was requested for an old version of the descriptor,
+				// return it even if there is only a short time left before it
+				// expires. We can't renew this lease as doing so would violate the
+				// invariant that we only get leases on the newest version. The
+				// transaction will either finish before the lease expires or it will
+				// abort, which is what will happen if we returned an error here.
+				s.refcount++
+				return s, nil
+			}
+			if s.expiration-store.clock.Now().WallTime >= minLeaseDuration {
+				s.refcount++
+				return s, nil
+			}
+		} else if version != 0 {
+			n := t.active.findNewest(0)
+			if n != nil && version < n.Version {
+				return nil, util.Errorf("table %d unable to acquire lease on old version: %d < %d",
+					t.id, version, n.Version)
+			}
+		}
+
+		if t.acquiring != nil {
+			// There is already a lease acquisition in progress. Wait for it to complete.
+			t.mu.Unlock()
+			<-t.acquiring
+			t.mu.Lock()
+		} else {
+			// There is no active lease acquisition so we'll go ahead and perform
+			// one.
+			t.acquiring = make(chan struct{})
+			t.mu.Unlock()
+			s, err := store.Acquire(t.id, version)
+			t.mu.Lock()
+			close(t.acquiring)
+			t.acquiring = nil
+			if err != nil {
+				return nil, err
+			}
+			t.active.insert(s)
+		}
+
+		// A new lease was added, so loop and perform the lookup again.
+	}
+}
+
+func (t *tableState) release(lease *LeaseState, store LeaseStore) error {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	s := t.active.find(lease.Version, lease.expiration)
+	if s == nil {
+		return util.Errorf("table %d version %d not found", lease.ID, lease.Version)
+	}
+	s.refcount--
+	if s.refcount == 0 {
+		n := t.active.findNewest(0)
+		if s != n {
+			if s.Version < n.Version {
+				// TODO(pmattis): If an active transaction is releasing the lease for
+				// an older version, hold on to it for a few seconds in anticipation of
+				// another operation being performed within the transaction. If we
+				// release the lease immediately the transaction will necessarily abort
+				// on the next operation due to not being able to get the lease.
+			}
+			t.active.remove(s)
+			return store.Release(s)
+		}
+	}
+	return nil
+}
+
+// LeaseManager manages acquiring and releasing per-table leases. Exported only
+// for testing.
+type LeaseManager struct {
+	LeaseStore
+	mu     sync.Mutex
+	tables map[ID]*tableState
+}
+
+// NewLeaseManager creates a new LeaseManager.
+func NewLeaseManager(nodeID uint32, db client.DB, clock *hlc.Clock) *LeaseManager {
```

Comment:
> A pointer signifies the internal state is mutated.

It certainly signifies that when it's a function receiver, but I don't see why that's true of a return value. Also, isn't it the case that there will be just one LeaseManager in the system which is owned by the one sql.executor in the system? Why do you need to share anything with anyone?